### ISSUE TITLE
Regression - ODataConventionModelBuilder.GetEdmModel Stack Overflow + Recursive Complex Types

### ISSUE THREAD

**Issue by timward60** [OData/WebApi#930]
I am seeing an stack overflow exception when my service tries starts up and returns the EdmModel using the ODataConventionModelBuilder. I believe this is because I have a model with entity that has a recursive complex type property.

```C#
public static IEdmModel GetConventionModel()    
{  
    var builder = new ODataConventionModelBuilder();  
    builder.EntitySet<Rule>("Rules");  
    builder.EntitySet<RuleVariable>("RuleVariables");  
    return builder.GetEdmModel(); // Stack overlfow thrown here.
}
```

I don't see this issue when using older version of System.Web.OData nad Microsoft.OData.*. Sample project with recursive complex types working can be found [here].(https://github.com/timward60/RuleService/tree/microsoft_odata_client_issue)

### Assemblies affected
System.Web.OData 6.0.0.0
Microsoft.OData.Core 7.0.0
Microsoft.OData.Edm 7.0.0

### Reproduce steps
1. Create model with an entity that contains a recursive complex type.

I also have a sample project that reproduces the bug:
> 1. git clone https://github.com/timward60/RuleService.git
> 2. git checkout tags/system_web_data_issue -b hotfix/system_web_data_issue
> 3. start RuleService.sln
> 4. Launch RuleService

### Expected result
1. GetEdmModel succeeds.

### Actual result
Stack overflow

### Additional details
[Example Metadata](https://gist.github.com/timward60/32ef54d706df8fed0324a041bed7569c#file-ruleservice_metadata-edmx)
[Full Exception](https://gist.github.com/timward60/32ef54d706df8fed0324a041bed7569c#file-ruleservice_microsoft_odata_core_exception-txt)
[Sample Project](https://github.com/timward60/RuleService/tree/system_web_data_issue)

**Comment by brjohnstmsft** [comment]
@robward-ms This issue is a blocker for Complex Types support in Azure Search. Seems to be a duplicate of https://github.com/OData/WebApi/issues/437  I'm also curious why the referenced PR was closed. If it solves the issue, it would be great to see it merged.

**Comment by robward-ms** [comment]
Looks like this was fixed in #977 and is available in WebApi 6.1.

**Comment by brjohnstmsft** [comment]
This was not completely fixed. See this issue for broader context: https://github.com/OData/WebApi/issues/1248
